runtime/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt

```diff
@@ -59,11 +61,17 @@ public open class PluginGeneratedSerialDescriptor(
 
     override fun getElementDescriptor(index: Int): SerialDescriptor {
         return generatedSerializer?.childSerializers()?.get(index)?.descriptor
-        ?: throw IndexOutOfBoundsException("$serialName descriptor has only $elementsCount elements, index: $index")
+                ?: throw IndexOutOfBoundsException("$serialName descriptor has only $elementsCount elements, index: $index")
+    }
+
+    internal val typeParameterDescriptors: List<SerialDescriptor> by lazy {
```

Comment:
1) `private`
2) It doesn't have to be lazy
3) `val` in the middle of the class
4) Better make it an array

Reply:
It has to be lazy because of native freezing (see my previous comment) and recursiveness: simply `generatedSerializer?.typeParametersSerializers()?.map { it.descriptor }` does not work if `descriptor` is the same descriptor being created (because it is not written to the field yet). Yeah, I know, recursive generics are not used often, but still.

It is also possible to convert it to the property with getter; it shouldn't hit the performance much, probably.

Reply:
Why the array is better here?

Reply:
Array simply has a lesser footprint

Reply:
I've added test for recursive generics and it passes even without lazy. However, JS test fail completely due to initialization order problem

Reply:
Because for some reason, in JS descriptor is created before actual assigment of type parameter serializers to fields.

Reply:
I'll leave `lazy` for now and try to fix #789 later